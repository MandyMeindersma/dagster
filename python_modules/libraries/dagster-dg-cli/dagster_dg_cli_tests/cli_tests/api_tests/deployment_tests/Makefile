.PHONY: rerecord
rerecord:
	dagster-dev dg-api-record deployment
