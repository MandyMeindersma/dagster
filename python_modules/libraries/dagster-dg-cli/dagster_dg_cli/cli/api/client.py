"""Client factory for DG API commands."""

from typing import Protocol

import click
from dagster_shared.plus.config import DagsterPlusCliConfig

from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient


class GraphQLClientFactory(Protocol):
    """Protocol for GraphQL client factories used in testing."""

    def __call__(self, config: DagsterPlusCliConfig) -> DagsterPlusGraphQLClient: ...


class DgApiTestContext:
    """Test context for DG API commands."""

    def __init__(self, client_factory: GraphQLClientFactory):
        self.client_factory = client_factory


def create_dg_api_graphql_client(
    ctx: click.Context, config: DagsterPlusCliConfig
) -> DagsterPlusGraphQLClient:
    """Create GraphQL client for DG API commands.

    This is the single entry point for GraphQL client creation in DG API commands.
    It checks for test context injection and falls back to creating a real client.

    Args:
        ctx: Click context from the CLI command
        config: Optional config to use, falls back to global config if not provided

    Returns:
        DagsterPlusGraphQLClient instance
    """
    # Check if we have a test context with custom factory
    if ctx.obj and isinstance(ctx.obj, DgApiTestContext) and ctx.obj.client_factory:
        return ctx.obj.client_factory(config)

    # Normal operation: create real client from config
    return DagsterPlusGraphQLClient.from_config(config)
