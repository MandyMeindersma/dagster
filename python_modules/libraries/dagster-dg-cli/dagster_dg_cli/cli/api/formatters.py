"""Output formatters for CLI display."""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dagster_dg_cli.dagster_plus_api.schemas.asset import DgPlusApiAsset, DgPlusApiAssetList
    from dagster_dg_cli.dagster_plus_api.schemas.deployment import DeploymentList


def format_deployments(deployments: "DeploymentList", as_json: bool) -> str:
    """Format deployment list for output."""
    if as_json:
        return deployments.model_dump_json(indent=2)

    lines = []
    for deployment in deployments.items:
        lines.extend(
            [
                f"Name: {deployment.name}",
                f"ID: {deployment.id}",
                f"Type: {deployment.type.value}",
                "",  # Empty line between deployments
            ]
        )

    return "\n".join(lines).rstrip()  # Remove trailing empty line


def format_assets(assets: "DgPlusApiAssetList", as_json: bool) -> str:
    """Format asset list for output."""
    if as_json:
        return assets.model_dump_json(indent=2)

    lines = []

    # Handle empty asset list
    if not assets.items:
        lines.append("No assets found.")
    else:
        for asset in assets.items:
            lines.extend(
                [
                    f"Asset Key: {asset.asset_key}",
                    f"ID: {asset.id}",
                    f"Group: {asset.group_name}",
                    f"Description: {asset.description or '(none)'}",
                    f"Kinds: {', '.join(asset.kinds) if asset.kinds else '(none)'}",
                    "",  # Empty line between assets
                ]
            )

    # Add pagination info
    if assets.cursor or assets.has_more:
        lines.extend(
            [
                "Pagination:",
                f"  Has More: {assets.has_more}",
            ]
        )
        if assets.cursor:
            lines.append(f"  Next Cursor: {assets.cursor}")
        lines.append("")

    return "\n".join(lines).rstrip()  # Remove trailing empty line


def format_asset(asset: "DgPlusApiAsset", as_json: bool) -> str:
    """Format single asset for output."""
    if as_json:
        return asset.model_dump_json(indent=2)

    lines = [
        f"Asset Key: {asset.asset_key}",
        f"ID: {asset.id}",
        f"Group: {asset.group_name}",
        f"Description: {asset.description or '(none)'}",
        f"Kinds: {', '.join(asset.kinds) if asset.kinds else '(none)'}",
    ]

    if asset.metadata_entries:
        lines.extend(["", "Metadata:"])
        for entry in asset.metadata_entries:
            label = entry.get("label", "")
            description = entry.get("description", "")
            lines.append(f"  {label}: {description}")

    return "\n".join(lines)
