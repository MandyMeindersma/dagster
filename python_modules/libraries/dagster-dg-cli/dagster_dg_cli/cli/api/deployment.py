"""Deployment API commands following GitHub CLI patterns."""

import sys

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper

from dagster_dg_cli.cli.api.formatters import format_deployments
from dagster_dg_cli.cli.api.shared import format_error_for_output, get_config_or_error
from dagster_dg_cli.dagster_plus_api.api.deployments import DgApiDeploymentApi


@click.command(name="list", cls=DgClickCommand, unlaunched=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@cli_telemetry_wrapper
def list_deployments_command(output_json: bool) -> None:
    """List all deployments in the organization."""
    config = get_config_or_error()
    api = DgApiDeploymentApi(config)

    try:
        deployments = api.list_deployments()
        output = format_deployments(deployments, as_json=output_json)
        click.echo(output)
    except Exception as e:
        error_output, exit_code = format_error_for_output(e, output_json)
        click.echo(error_output, err=True)
        sys.exit(exit_code)


@click.group(
    name="deployment",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "list": list_deployments_command,
    },
)
def deployment_group():
    """Manage deployments in Dagster Plus."""
