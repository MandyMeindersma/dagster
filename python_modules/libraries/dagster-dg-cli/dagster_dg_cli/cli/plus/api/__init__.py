"""API commands for interacting with Dagster Plus REST-like interface."""

import click
from dagster_dg_core.utils import DgClickGroup

from dagster_dg_cli.cli.plus.api.deployments import deployments_group


@click.group(
    name="api",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "deployments": deployments_group,
    },
)
def api_group():
    """Make REST-like API calls to Dagster Plus."""
