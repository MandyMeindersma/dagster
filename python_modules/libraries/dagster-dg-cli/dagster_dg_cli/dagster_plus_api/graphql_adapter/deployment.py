"""GraphQL implementation for deployment operations."""

from typing import TYPE_CHECKING, Any, Optional

from dagster_shared.plus.config import DagsterPlusCliConfig

from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.dagster_plus_api.schemas.deployment import DeploymentList

# GraphQL queries
LIST_DEPLOYMENTS_QUERY = """
query ListDeployments {
    fullDeployments {
        deploymentName
        deploymentId
        deploymentType
    }
}
"""


def process_deployments_response(
    graphql_response: dict[str, Any], limit: Optional[int] = None
) -> "DeploymentList":
    """Process GraphQL response into DeploymentList.

    This is a pure function that can be easily tested without mocking GraphQL clients.

    Args:
        graphql_response: Raw GraphQL response containing "fullDeployments"
        limit: Optional limit to apply to results

    Returns:
        DeploymentList: Processed deployment data
    """
    # Import pydantic models only when needed
    from dagster_dg_cli.dagster_plus_api.schemas.deployment import (
        Deployment,
        DeploymentList,
        DeploymentType,
    )

    deployments_data = graphql_response.get("fullDeployments", [])

    deployments = [
        Deployment(
            id=d["deploymentId"],
            name=d["deploymentName"],
            type=DeploymentType[d["deploymentType"]],
        )
        for d in deployments_data
    ]

    # Apply limit if specified - GraphQL fullDeployments does not support limit parameter
    # so we must filter client-side
    if limit:
        deployments = deployments[:limit]

    return DeploymentList(
        items=deployments,
        total=len(deployments),
    )


def list_deployments_via_graphql(
    config: DagsterPlusCliConfig,
    limit: Optional[int] = None,
) -> "DeploymentList":
    """Fetch deployments using GraphQL.
    This is an implementation detail that can be replaced with REST calls later.
    """
    client = DagsterPlusGraphQLClient.from_config(config)
    result = client.execute(LIST_DEPLOYMENTS_QUERY)

    return process_deployments_response(result, limit=limit)
