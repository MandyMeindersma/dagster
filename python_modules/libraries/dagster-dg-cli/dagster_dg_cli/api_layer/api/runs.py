"""Run endpoints - REST-like interface."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.run import get_run_via_graphql, list_runs_via_graphql
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.run import Run, RunList


@dataclass(frozen=True)
class DgApiRunApi:
    client: IGraphQLClient

    def list_runs(
        self,
        limit: Optional[int] = None,
        status: Optional[str] = None,
        job_name: Optional[str] = None,
        cursor: Optional[str] = None,
    ) -> "RunList":
        """List runs with optional filtering.

        Args:
            limit: Maximum number of runs to return
            status: Filter by run status
            job_name: Filter by job name
            cursor: Pagination cursor

        Returns:
            RunList: List of runs matching the filters
        """
        return list_runs_via_graphql(
            client=self.client,
            limit=limit,
            status=status,
            job_name=job_name,
            cursor=cursor,
        )

    def get_run(self, run_id: str) -> "Run":
        """Get a specific run by ID.

        Args:
            run_id: The run ID to fetch

        Returns:
            Run: The requested run

        Raises:
            Exception: If run not found or API error occurs
        """
        return get_run_via_graphql(client=self.client, run_id=run_id)
