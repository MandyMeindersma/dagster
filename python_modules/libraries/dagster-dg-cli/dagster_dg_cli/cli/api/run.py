"""Run API commands following GitHub CLI patterns."""

import json
from typing import Optional

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper
from dagster_shared.plus.config import DagsterPlusCliConfig
from dagster_shared.plus.config_utils import dg_api_options

# Lazy import to avoid loading pydantic at CLI startup
from dagster_dg_cli.cli.api.client import create_dg_api_graphql_client
from dagster_dg_cli.cli.api.formatters import format_run, format_runs


@click.command(name="list", cls=DgClickCommand, unlaunched=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@click.option(
    "--status",
    help="Filter by run status (e.g., SUCCESS, FAILURE, STARTED)",
)
@click.option(
    "--job",
    "job_name",
    help="Filter by job name",
)
@click.option(
    "--limit",
    type=int,
    help="Maximum number of runs to return",
)
@dg_api_options(deployment_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def list_runs_command(
    ctx: click.Context,
    output_json: bool,
    status: Optional[str],
    job_name: Optional[str],
    limit: Optional[int],
    organization: str,
    deployment: str,
    api_token: str,
    view_graphql: bool,
) -> None:
    """List runs in the deployment."""
    config = DagsterPlusCliConfig.create_for_deployment(
        organization=organization,
        deployment=deployment,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    from dagster_dg_cli.api_layer.api.runs import DgApiRunApi

    api = DgApiRunApi(client)

    try:
        runs = api.list_runs(
            limit=limit,
            status=status,
            job_name=job_name,
        )
        output = format_runs(runs, as_json=output_json)
        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)
        raise click.ClickException(f"Failed to list runs: {e}")


@click.command(name="get", cls=DgClickCommand, unlaunched=True)
@click.argument("run_id", required=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@dg_api_options(deployment_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def get_run_command(
    ctx: click.Context,
    run_id: str,
    output_json: bool,
    organization: str,
    deployment: str,
    api_token: str,
    view_graphql: bool,
) -> None:
    """Get detailed information about a specific run."""
    config = DagsterPlusCliConfig.create_for_deployment(
        organization=organization,
        deployment=deployment,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config, view_graphql=view_graphql)
    from dagster_dg_cli.api_layer.api.runs import DgApiRunApi

    api = DgApiRunApi(client)

    try:
        run = api.get_run(run_id)
        output = format_run(run, as_json=output_json)
        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)
        raise click.ClickException(f"Failed to get run: {e}")


@click.group(
    name="run",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "list": list_runs_command,
        "get": get_run_command,
    },
)
def run_group():
    """Manage runs in Dagster Plus."""
