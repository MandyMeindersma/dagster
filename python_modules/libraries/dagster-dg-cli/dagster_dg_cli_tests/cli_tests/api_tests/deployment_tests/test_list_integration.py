"""Integration tests for deployment list command.

These tests focus on CLI behavior and use minimal mocking of the GraphQL client
to test the integration between components.
"""

import json
from unittest.mock import patch

from click.testing import CliRunner
from dagster_dg_cli.cli.api.deployment import list_deployments_command

from dagster_dg_cli_tests.cli_tests.api_tests.deployment_tests.fixtures import (
    load_deployment_response_fixture,
)


class TestDeploymentListCommand:
    """Test the deployment list CLI command integration."""

    @patch("dagster_dg_cli.cli.api.deployment._get_config_or_error")
    @patch(
        "dagster_dg_cli.dagster_plus_api.graphql_adapter.deployment.DagsterPlusGraphQLClient.from_config"
    )
    def test_list_deployments_text_output(self, mock_client_class, mock_config, snapshot):
        """Test successful deployment list with text output."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_client = mock_client_class.return_value
        mock_client.execute.return_value = load_deployment_response_fixture(
            "success_multiple_deployments"
        )

        # Run command
        runner = CliRunner()
        result = runner.invoke(list_deployments_command, [])

        # Verify success
        assert result.exit_code == 0

        # Verify GraphQL was called with correct query
        mock_client.execute.assert_called_once()
        called_query = mock_client.execute.call_args[0][0]
        assert "query ListDeployments" in called_query
        assert "fullDeployments" in called_query

        # Snapshot the CLI output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.deployment._get_config_or_error")
    @patch(
        "dagster_dg_cli.dagster_plus_api.graphql_adapter.deployment.DagsterPlusGraphQLClient.from_config"
    )
    def test_list_deployments_json_output(self, mock_client_class, mock_config, snapshot):
        """Test successful deployment list with JSON output."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_client = mock_client_class.return_value
        mock_client.execute.return_value = load_deployment_response_fixture(
            "success_multiple_deployments"
        )

        # Run command with --json flag
        runner = CliRunner()
        result = runner.invoke(list_deployments_command, ["--json"])

        # Verify success
        assert result.exit_code == 0

        # Parse and snapshot the JSON output to avoid formatting differences
        actual_output = json.loads(result.output)
        snapshot.assert_match(actual_output)

    @patch("dagster_dg_cli.cli.api.deployment._get_config_or_error")
    @patch(
        "dagster_dg_cli.dagster_plus_api.graphql_adapter.deployment.DagsterPlusGraphQLClient.from_config"
    )
    def test_list_deployments_api_error(self, mock_client_class, mock_config, snapshot):
        """Test handling of API errors."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_client = mock_client_class.return_value
        mock_client.execute.side_effect = Exception("API Error")

        # Run command
        runner = CliRunner()
        result = runner.invoke(list_deployments_command, [])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.deployment._get_config_or_error")
    @patch(
        "dagster_dg_cli.dagster_plus_api.graphql_adapter.deployment.DagsterPlusGraphQLClient.from_config"
    )
    def test_list_deployments_api_error_json_output(self, mock_client_class, mock_config, snapshot):
        """Test handling of API errors with JSON output."""
        # Setup mocks
        mock_config.return_value = "mock-config"
        mock_client = mock_client_class.return_value
        mock_client.execute.side_effect = Exception("API Error")

        # Run command with --json flag
        runner = CliRunner()
        result = runner.invoke(list_deployments_command, ["--json"])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output
        snapshot.assert_match(result.output)

    @patch("dagster_dg_cli.cli.api.deployment.DagsterPlusCliConfig.exists")
    def test_list_deployments_no_config(self, mock_config_exists, snapshot):
        """Test handling when no Dagster Plus config exists."""
        # Setup mock - no config exists
        mock_config_exists.return_value = False

        # Run command
        runner = CliRunner()
        result = runner.invoke(list_deployments_command, [])

        # Verify error handling
        assert result.exit_code != 0

        # Snapshot the error output
        snapshot.assert_match(result.output)
