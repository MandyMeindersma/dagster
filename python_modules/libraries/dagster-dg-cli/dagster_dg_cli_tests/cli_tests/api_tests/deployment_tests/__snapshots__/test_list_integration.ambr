# serializer version: 1
# name: TestDeploymentListCommand.test_list_deployments_api_error[0]
  '''
  Error querying Dagster Plus API: API Error
  Error: Failed to list deployments: API Error
  
  '''
# ---
# name: TestDeploymentListCommand.test_list_deployments_api_error_json_output[0]
  '''
  {"error": "API Error"}
  Error: Failed to list deployments: API Error
  
  '''
# ---
# name: TestDeploymentListCommand.test_list_deployments_json_output[0]
  dict({
    'items': list([
      dict({
        'id': 6935,
        'name': 'uat',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6936,
        'name': 'excessive',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 84915,
        'name': 'ben-test',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 37,
        'name': 'ecs',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6934,
        'name': 'staging',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 80,
        'name': 'prod',
        'type': 'PRODUCTION',
      }),
    ]),
    'total': 6,
  })
# ---
# name: TestDeploymentListCommand.test_list_deployments_no_config[0]
  '''
  Usage: list [OPTIONS]
  Try 'list --help' for help.
  
  Error: `dg plus` commands require authentication with Dagster Plus. Run `dg plus login` to authenticate.
  
  '''
# ---
# name: TestDeploymentListCommand.test_list_deployments_text_output[0]
  '''
  Name: uat
  ID: 6935
  Type: PRODUCTION
  
  Name: excessive
  ID: 6936
  Type: PRODUCTION
  
  Name: ben-test
  ID: 84915
  Type: PRODUCTION
  
  Name: ecs
  ID: 37
  Type: PRODUCTION
  
  Name: staging
  ID: 6934
  Type: PRODUCTION
  
  Name: prod
  ID: 80
  Type: PRODUCTION
  
  '''
# ---
